server/user.go

```diff
@@ -164,3 +164,46 @@ func uploadUserAvatar(c *gin.Context, db *gorm.DB, userId uint) (Image, error) {
 	}
 	return img, nil
 }
+
+func userChangePassword(db *gorm.DB, pwds UserPasswordUpdateRequest, userId uint) error {
+	slog.Debug("userChangePassword request running", "user_id", userId)
+	user := new(User)
+	res := db.Where("id = ?", userId).Take(&user)
+	if res.Error != nil {
+		slog.Error("userChangePassword failed - failed to retrieve user from database", "user_id", userId, "error", res.Error)
+		return errors.New("failed to retrieve user")
+	} else {
+		slog.Debug("userChangePassword user found", "user_id", userId)
+		match, err := compareHash(pwds.OldPassword, user.Password)
+		if err != nil {
+			slog.Error("userChangePassword failed - failed to compare passwords", "user_id", userId, "error", err)
+			return errors.New("failed to compare passwords")
+		}
+		if !match {
+			slog.Error("userChangePassword failed - current password hash doesn't match password hash in database", "user_id", userId, "error", err)
+			return errors.New("current password provided doesn't match password in database")
+		}
+		slog.Debug("userChangePassword hash for current password matches hash in the database", "user_id", userId)
+		slog.Debug("userChangePassword hashing new password", "user_id", userId)
+		hash, err := hashPassword(pwds.NewPassword, &ArgonParams{
```

Comment:
This is up to you if you want to. Since the `ArgonParams` must be the same here as in the `register` function, we could move them both out to a `func` (in `auth.go`):

```
func GetPassArgonParams() *ArgonParams {
	return &ArgonParams{
		memory:      64 * 1024,
		iterations:  3,
		parallelism: 2,
		saltLength:  16,
		keyLength:   32,
	}
}
```

I'm leaning towards a func (instead of a var) so we can rest easy knowing the parameters can't be changed (for everything).

Reply:
Totally agree in `ArgonParams` being a `func` instead of `var`!